### PR TITLE
108 gene pool test

### PR DESCRIPTION
**Pull request by KamilKamilS**
Not a final version

### CODE REVIEW

**Comment by PawelDabrowski83** [comment]
Tests for Mutator and Evaluator seems like integration testing, not a unit tests. I wonder is there a way to clearly assume that Mutator is working. In our scenario we should not assume that mutationChance is 100%, because this would be very unusual case.

What do you think about something like this: we look at Mutator's mock and check how many times it calls method mutate(). If I am not mistaken, mutator.mutate(Gene) should be calles n times, where n is poolOfGene.size().
Of course you may ask, how do we know if mutate() was called on each Gene once and n-times on first Gene from poolOfGene. I do not have answer for this, but I try to look for something.
Maybe there is a way to fill poolOfGene with mocks just to check if every item was passed?

**Comment by PawelDabrowski83** [comment]
I think that test are good, but somehow they rely on implementation of other classes, ie. to test Mutator we assume that mutationChance should be 1F to assure transformation. 
But if we someday rebuild Mutator class completely, just leaving mutate() intact, as this is part of interface - then we should look closer and modify unit tests for GenePool, which feels just unnecessary.

**Review by PawelDabrowski83 — APPROVED** [review]

File: src/test/java/com/facebook/genAlgo/genepool/GenePoolTest.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+    @Mock
+    RandomProvider randomProvider;
+
+    @Mock
+    MutatorService mutatorService;
+
+    @Mock
+    Evaluator evaluator;
+
```
Comment (PawelDabrowski83) [inline src/test/java/com/facebook/genAlgo/genepool/GenePoolTest.java:28]: I recommend to add @DisplayName with short description of each test

File: src/test/java/com/facebook/genAlgo/genepool/GenePoolTest.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        // when
+        genePool.evaluateFitness();
+
+        // then
+        verify(evaluator, times(sizeExpected)).setFitness(anyObject());
+    }
+
+    @Test
+    @ValueSource(ints = {2, 10, 30, 55, 1000})
+    public void shouldPerformEvolution(int sizeExpected) {
```
Comment (PawelDabrowski83) [inline src/test/java/com/facebook/genAlgo/genepool/GenePoolTest.java:70]: I suggest to change this test to: shouldIncreaseGenerationWhenPerformEvolution and use setGeneration to number from ValueSource.
assert that generation is increased by 1 after performEvolution.

**Review by PawelDabrowski83 — APPROVED** [review]

**Comment by PawelDabrowski83** [comment]
performEvolution() does 3 things: mutate, evaluate and increase generation, so we should split every of this function to another test/assertion.

test for makeMutation() checks if mutator is called n times, but does not check if every gene in pool is modified, so it does not cover all use cases. I could pass it mutating first gene n times.

**Comment by KamilKamilS** [comment]
test changed, according to your ideas.
I agreed that we do not check that mutate(gene) method takes each time different argument. We need to check it, I was looking for a smart solution how to check it, but not found anything yet.

**Pull request merged**